OCMUI-3196: Consolidate all 'access.redhat.com' links into single file

dtaylor113 opened the pull request:
# Summary

**Consolidate access.redhat.com URLs links into a single file**

This PR consolidates all hardcoded `access.redhat.com` URLs from React components into a new `supportLinks.mjs` file.

Additionally, this PR implements DRY principles by:
- Creating a consistent `getLinks()` interface across all link files
- Centralizing URL combination logic in `combineAndSortLinks()` utility
- Establishing a single `getAllExternalLinks()` function for scripts to use

**Assisted by: Cursor/claude-3.5-sonnet**

# Jira

Addresses [OCMUI-3196](https://issues.redhat.com/browse/OCMUI-3196)

# Additional Information

## Updated Components (19 total)

All components updated with 12 constants from `supportLinks.mjs`:

**Support Case Management (4 components):**
- `ClusterStatusMonitor.jsx` → `SUPPORT_CASE_NEW`
- `ClusterListTable.jsx` → `SUPPORT_CASE_NEW` 
- `AddOnsDrawerPrimaryButton.jsx` → `SUPPORT_CASE_NEW`
- `AddOnsDrawerFailedBox.jsx` → `SUPPORT_CASE_NEW`
- `SupportCasesCardHelper.tsx` → `SUPPORT_CASE_NEW_WITH_ISSUE` + `SUPPORT_CASE_VIEW`

**Support Home Page (3 components):**
- `ExpirationAlert.tsx` → `SUPPORT_HOME`
- `TokenErrorAlert.jsx` → `SUPPORT_HOME`
- `SupportLevelLabel.tsx` → `SUPPORT_HOME`

**Knowledge Base Articles (5 components):**
- `errors.ts` → `EXPORT_CONTROL_KB`
- `OfflineTokensAlert.tsx` → `OFFLINE_TOKENS_KB`
- `TransferClusterOwnershipDialog.tsx` → `PULL_SECRET_CHANGE_KB`
- `TransferClusterOwnershipInfo.tsx` → `PULL_SECRET_CHANGE_KB`
- `BillingModel.tsx` → `BILLING_MODEL_KB`
- `ArchiveClusterDialog.tsx` → `ARCHIVE_CLUSTER_KB`

**Support Policies (4 components):**
- `SupportStatus.tsx` → `OPENSHIFT_SUPPORT_POLICY`
- `ClusterCreatedIndicator.jsx` → `OPENSHIFT_POLICY_UPDATES`
- `UpgradeSettingsFields.jsx` (3 files) → `SECURITY_CLASSIFICATION_CRITICAL`
- `ClusterUpdates.tsx` → `SECURITY_CLASSIFICATION_CRITICAL`

## Link Checker Results

./run/check-links.mjs

**Before Changes (baseline):**
- Total URLs: 402 (from installLinks.mjs only)

**After Changes:**
- Total URLs: 414 (+12 new support URLs)
- All URLs accessible ✅
- No 404 errors ✅

## Architecture & DRY Improvements

**New Files Created:**
- `src/common/supportLinks.mjs` - Centralized support URLs
- `src/common/urlUtils.mjs` - Shared URL utilities
- `src/common/__tests__/urlConsolidation.test.jsx` - Consolidation tests urls match urls before this PR.

**DRY Implementation:**
- **Consistent Interface**: All link files export `getLinks()` 
- **Hidden Complexity**: `installLinks.mjs` flattening logic is internal
- **Centralized Scripts**: Single `getAllExternalLinks()` for all scripts
- **Shared Utilities**: `combineAndSortLinks()` eliminates duplication

## Implementation/Architectural Decisions

**Two-Phase Approach**: This PR consolidates `access.redhat.com` URLs into a new supportLinks.mjs file; a follow-up ticket [OCMUI-3744](https://issues.redhat.com/browse/OCMUI-3744) will redistribute the links in `installLinks.mjs`, and move some to supportLinks.mjs, and other to a new docLinks.mjs files. 

# How to Test

## Automated Testing

**Unit Tests (21 tests added):**

The new urlConsolidation.test tests that each component displays the exact url/link it did before this PR.  

🔴 IMPORTANT: **_The urlConsolidation.test is verification the consolidation using the new constants, is rendering the same exact urls prior this PR/consolidation  -It is a 'comfort' check just for this PR -that all urls remain as before, as such, it won't be merged into main.  It is just for Reviewers to run locally to test the urls if they want_**

```bash
yarn test src/common/__tests__/urlConsolidation.test.jsx
```
- Tests verify each migrated component renders same URL as pre-consolidation
- Tests verify `getAllExternalLinks()` utility returns sorted, unique URLs
- Tests verify `overrideErrorMessage()` function returns same message format

**Link Validation:**
```bash
yarn run check-links    # Validates all 414 URLs are accessible
yarn run dump-links     # Lists all collected URLs
```

## Manual Testing

**Verify Links Render Correctly:**

1. **Billing Information**:
   - Navigate to OSD Wizard, Billing Step → "Try OpenShift Dedicated" should link to billing KB
2. **Archive Cluster**:
   - Archive a cluster, the "Instructions" link in the dialog should link to a kb article
3. **/openshift/token**:
   - "Additional Resources" the link should link to a kb article
   
**Visual Verification**: All "Learn more" and "Contact support" links should point to `access.redhat.com` URLs and open in new tabs.

# Review process

Please review and follow the [PR process](https://github.com/RedHatInsights/uhc-portal/blob/main/docs/pull-request-process.md).

## QE Reviewer

- [x] _Pre-merge testing : Verified change locally in a browser (downloaded and ran code using reviewx tool)_
- [x] Updated/created Polarion test cases which were peer QE reviewed
- [x] Confirmed 'tc-approved' label was added by dev to the linked JIRA ticket
- [ ] (optional) Updated/created Cypress e2e tests
- [ ] Closed threads I started after the author made changes or added an explanation

dtaylor113 commented:
cc @kdoberst who recently hosted the ./run/check-links.mjs script via a GitHub action.  
This PR should not change/impact that at all.

LiorKGOW reviewed (commented):
Thank you for the PR @dtaylor113 
I am still reviewing it, but wanted to give you some things I found

dtaylor113 commented on src/common/linkUtils.mjs:
```diff
@@ -0,0 +1,12 @@
+ * @param {...Array<string>} linkSources - Arrays of links to combine
+ * @returns {Array<string>} - Sorted array of unique links
+ */
+export const combineAndSortLinks = (...linkSources) => {
+  const allLinks = linkSources.flat();
+  return [...new Set(allLinks)].sort();
```
Hi, yes, that is the case

eliranmal commented on src/common/linkUtils.mjs:
```diff
@@ -0,0 +1,14 @@
+ * @param {...Array<string>} linkSources - Arrays of links to combine
+ * @returns {Array<string>} - Sorted array of unique links
+ */
+export const combineAndSortLinks = (...linkSources) => {
```
IMO that function is generic enough to just go in _src/common/helpers.ts_ (it seems to hold other array-related utils)

dtaylor113 replied:
Hi @eliranmal, I did move it into 'src/common/helpers.ts' locally and got:

"I can see the dependency cycle issue. The combineAndSortLinks function doesn't actually need any of the complex imports in helpers.ts, but since helpers.ts imports from userReducer which imports from errors, it creates a cycle.
The cleanest solution is to create a separate file for link-related utilities. Let me create linkHelpers.ts:"

...and it basically reverted what I had done.  :-)

eliranmal replied:
also, how about just putting in inside 'installLinks'?

eliranmal replied:
that's it - i'm pulling my hands from this thread cuz i'm not even supposed to review this!

dtaylor113 commented on package.json:
```diff
@@ -223,6 +223,7 @@
     "insights-proxy": "make insights-proxy-check && run/insights-proxy.sh",
     "stop-insights-proxy": "run/podman-or-docker.sh kill insightsproxy; run/podman-or-docker.sh rm --force insightsproxy; true",
     "lint": "eslint src 'run/*js' --ext mjs,js,jsx,ts,tsx",
+    "lint:files": "eslint --ext mjs,js,jsx,ts,tsx",
```
"lint" by itself does "... eslint src 'run/*js'", has always 'run all the files under root'.
There is currently no 'yarn lint' yarn script/cmd to target specific 'files'.
'lint:files' allows one to specify files to run against.

eliranmal replied:
that's it - i'll let you ponder my feedback and draw my hands from this thread, too (i'm not even supposed to review this PR).

dtaylor113 replied:
Hmm...I'm looking at .husky/pre-commit
I don't think this happens for me on my Mac?
I often do commits, push, and PRs, only to find out 'yarn lint' fails in CI because I forgot to run 'yarn lint' locally. 
!?

eliranmal replied:
if lint-staged is broken, we need to fix it.  still no reason to add another command

dtaylor113 replied:
Hi, reverted the 'lint:files' script

LiorKGOW commented on run/check-links.mjs:
```diff
@@ -62,9 +62,10 @@ Options:
-  URLs are obtained automatically from the getFlatUrls() function in
-  '../src/common/installLinks.mjs' which extracts URLs from the project.
+  URL Source:
+  URLs are obtained automatically from getAllExternalLinks() in '../src/common/urlUtils.mjs' which combines:
+  - '../src/common/installLinks.mjs' (installation and binary URLs)
+  - '../src/common/supportLinks.mjs' (support cases and knowledge base URLs)
```
There is no need to state where this function is imported from, we have the import on line 19 for that (DRY right 😉)

Also, we should follow the [core OOP principle of Abstraction](https://stackify.com/oop-concept-abstraction/#:~:text=Its%20main%20goal%20is%20to%20handle%20complexity%20by%20hiding%20unnecessary%20details%20from%20the%20user.) and keep it's implementation where it is defined. Keep it simple 😄

LiorKGOW reviewed (changes requested):
Thank you for the PR @dtaylor113 ! 
A few overall questions, in addition to the comments I had 😇 
1. What is the purpose of having a supportLinks & a instalLinks ? Both seem to have the same functions of exports and anyway we build a new "links" object right? Is there a reason to separate them ?
2. I see as lot of tests which have something wrong in their core. 
Tests should focus on the **current expected state of the code** not on what has changed on a specific PR.
I would advise adjusting them accordingly (like this one for example - link:   describe('overrideErrorMessage function', () => {
    it('should return same message as before consolidation for BANNED_USER_CODE', () => {
    ). 
link -> This is more accurate to what the test should test (related to testing the expected result more than the changes this PR brings) Just an example

dtaylor113 commented on src/common/__tests__/urlConsolidation.test.jsx:
```diff
@@ -0,0 +1,387 @@
+      // Get all URLs using the centralized function like our scripts do
+      const allUrls = await getAllExternalLinks();
+
+      // Should be an array
+      expect(Array.isArray(allUrls)).toBe(true);
```
Hi @LiorKGOW, sorry, I think you missed my comment in the PR's description:

"_The new urlConsolidation.test tests that each component displays the exact url/link it did before this PR. Just a verification the consolidation to using constants is working as expected.. It is a 'comfort' check just for this PR -that all urls remain as before, as such, I probably won't be merging it into main._"

There is no need to review this unit test, I will not be checking it in.

zherman0 reviewed (approved):
LGTM.  I know there are a lot of great on going conversations and I assume there will be changes to this, but I am fine with the current implementation as well.

eliranmal commented on run/dump-links.mjs:
```diff
@@ -1,6 +1,6 @@
-const flatUrls = await getFlatUrls();
-flatUrls.forEach((url) => console.log(url));
+const allUrls = await getAllExternalLinks();
+allUrls.forEach((url) => console.log(url));
```
git history search yields nothing; it's never been referenced anywhere, and probably created as a local tool or with the intention to be used in CI.

dtaylor113 commented on src/common/installLinks.mjs:
```diff
@@ -1010,14 +1012,17 @@ const urlsSelector = (githubReleases) => {
+  const nestedUrls = Object.values(urls)
+    .flatMap(Object.values)
+    .flatMap(Object.values)
+    .flatMap(Object.values);
```
Ok, changed to use `depth`, it's one line longer, and not sure it's more readable 🤷

dtaylor113 commented on src/common/supportLinks.mjs:
Nice!  Fixed.  Thanks for pointing this out, perfect review comment :-)

dtaylor113 commented on src/common/supportLinks.mjs:
```diff
@@ -0,0 +1,32 @@
+    'https://access.redhat.com/security/updates/classification/#critical',
+};
+
+/** Returns all support-related external links. */
```
Hi, removed.

dtaylor113 commented on src/common/__tests__/urlConsolidation.test.jsx:
```diff
@@ -0,0 +1,387 @@
+  });
+
+  describe('overrideErrorMessage function', () => {
+    it('should return same message as before consolidation for BANNED_USER_CODE', () => {
+      // Test the overrideErrorMessage function that was changed during consolidation
+      const payload = { code: BANNED_USER_CODE };
+      const result = overrideErrorMessage(payload);
+
+      // This is what the message should be with the hardcoded URL (pre-consolidation)
+      const expectedMessage = `Your account has been placed on Export Hold based on export control screening.
+The Export Compliance Team has been notified that your account is on hold, and must conduct additional due diligence to resolve the Export Hold.
+Try again in 24-48 hours.
+Learn more: https://access.redhat.com/articles/1340183`;
+
+      // Verify the function returns exactly the same message as before
+      expect(result).toBe(expectedMessage);
```
Ok, just to be clear, you should not test "src/common/tests/urlConsolidation.test.jsx", even after a rebase :-)
I will remove it before the final merge into main.

dtaylor113 commented:
Addressed @LiorKGOW 's comment:

"How about using the depth parameter the flat function of Arrays has?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat#depth
It would be more readable IMHO (DRY right? 😉)"

- fixed

dtaylor113 commented on src/common/__tests__/urlConsolidation.test.jsx:
Just to be clear, do not review [urlConsolidation.test.jsx](https://github.com/RedHatInsights/uhc-portal/pull/102/files#diff-6e1cefc42f97f5ea709d29c8a528f6058a6a38bec001baf4340ac6ef7303173b) it is just a 'comfort' unit test for these changes, it is not a long term unit test and thus will not be checked in with this review.

davidaznaur reviewed (approved):
Checked out locally. LGTM

dtaylor113 commented:
Hi @apecha, OSD Wizard, Billing Step → "Try OpenShift Dedicated" now opens in a new browser new tab.  thanks
